sync_diff_inspector/utils/utils.go

```diff
@@ -539,18 +531,18 @@ func CompareData(map1, map2 map[string]*dbutil.ColumnData, orderKeyCols, columns
 		str1 = string(data1.Data)
 		str2 = string(data2.Data)
 		if column.FieldType.GetType() == mysql.TypeFloat || column.FieldType.GetType() == mysql.TypeDouble {
-			if data1.IsNull == data2.IsNull && data1.IsNull {
-				continue
-			}
-
-			num1, err1 := strconv.ParseFloat(str1, 64)
-			num2, err2 := strconv.ParseFloat(str2, 64)
-			if err1 != nil || err2 != nil {
-				err = errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
-				return
-			}
-			if math.Abs(num1-num2) <= 1e-6 {
+			if data1.IsNull && data2.IsNull {
 				continue
+			} else if !data1.IsNull && !data2.IsNull {
+				num1, err1 := strconv.ParseFloat(str1, 64)
+				num2, err2 := strconv.ParseFloat(str2, 64)
+				if err1 != nil || err2 != nil {
+					err = errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
+					return
+				}
+				if math.Abs(num1-num2) <= 1e-6 {
```

Comment:
this hard coded threshold is not as good as 
```
		if col.FieldType.GetType() == mysql.TypeFloat {
			name = fmt.Sprintf("round(%s, 5-floor(log10(abs(%s)))) as %s", name, name, name)
		} else if col.FieldType.GetType() == mysql.TypeDouble {
			name = fmt.Sprintf("round(%s, 14-floor(log10(abs(%s)))) as %s", name, name, name)
		}
```

Reply:
We can use `num1=num2` to check it is equal or not. I want to know why float/double type need to be converted when comparing. Can you give me some examples?

When I canceled this code. It can also distinguish between source and target about float/double type.
https://github.com/pingcap/tidb-tools/blob/2218c401d5d7c0541d77fbc39adcd33d3dff2d81/sync_diff_inspector/utils/utils.go#L149-L153
https://github.com/pingcap/tidb-tools/blob/2218c401d5d7c0541d77fbc39adcd33d3dff2d81/sync_diff_inspector/utils/utils.go#L764-L768

Reply:
🤨  This is a history issue. Currently this 'hack' sub-expression `round(%s, 5-floor(log10(abs(%s))))` is done by Database itself, in other words, the database(TiDB/MySQL) will return the same result of `select (float value)` as the result of the sub-expression.
I think the version is high enough to remove this codes. 

> When I canceled this code. It can also distinguish between source and target about float/double type.
We also need to check whether the result of checksum is `false negative`, which means some checksum should be the same but is different, leads to trigger row compare operation(cost hightly).

Reply:
Got it. tks.